examples/Tutorial 4 - 3SAT Solver.ipynb

```diff
@@ -0,0 +1,477 @@
+    "    \"\"\"\n",
+    "\n",
+    "    terms: list[ThreeSATTerm]\n",
+    "    \n",
+    "    def __init__(self, terms : list[ThreeSATTerm]):\n",
+    "        if len(terms) != 3:\n",
+    "            raise ValueError(\"A 3SAT clause must have exactly three terms.\")\n",
+    "        self.terms = terms\n",
+    "\n",
+    "    def eval(self, values: dict[str, bool]) -> bool:\n",
+    "        \"\"\"\n",
+    "        Evaluate this clause with the value of the variables as specified in `values`.\n",
+    "        Returns `True` if and only for any `term`, `term.eval(values)` returns `True`.\n",
+    "        \"\"\"\n",
+    "        if len(self.terms) != 3:\n",
+    "            raise ValueError(\"A 3SAT clause must have exactly three terms.\")\n",
+    "        for term in self.terms:\n",
+    "            if term.eval(values):\n",
+    "                return True\n",
+    "        return False\n",
+    "    \n",
+    "    \n",
+    "\n",
+    "\n",
+    "class ThreeSATInstance:\n",
+    "    \"\"\"\n",
+    "    A 3SAT instance, i.e. C1 AND C2 AND C3  AND Cn where each Ci is a 3SAT Clause.\n",
+    "    \"\"\"\n",
+    "\n",
+    "    clauses: list[ThreeSATClause]\n",
+    "    _variables: set[str]\n",
+    "    \n",
+    "    def __init__(self, clauses : list[ThreeSATClause]):\n",
+    "        self.clauses = clauses\n",
+    "        self._variables = set()\n",
+    "        \n",
+    "        for clause in self.clauses:\n",
+    "            for term in clause.terms:\n",
+    "                self._variables.add(term.variable)\n",
+    "\n",
+    "\n",
+    "    def eval(self, values: dict[str, bool]) -> bool:\n",
+    "        \"\"\"\n",
+    "        Evaluate this instance with the value of the variables as specified in `values`.\n",
+    "\n",
+    "        Returns `True` if and only for each `clause`, `clause.eval(values)` returns `True`.\n",
+    "        \"\"\"\n",
+    "        for clause in self.clauses:\n",
+    "            if not clause.eval(values):\n",
+    "                return False\n",
+    "        return True\n",
+    "        \n",
+    "    \n",
+    "\n",
+    "    def __len__(self) -> int:\n",
+    "        \"\"\"\n",
+    "        The number of clauses in this instance.\n",
+    "        \"\"\"\n",
+    "        return len(self.clauses)\n",
+    "    \n",
+    "\n",
+    "    def compile_to_mis(self) -> MISInstance:\n",
+    "        \"\"\"\n",
+    "            Compile this instance of 3SAT into an instance of MIS.\n",
+    "        \"\"\"\n",
+    "        \n",
+    "        \"\"\"\n",
+    "        Making a dictionary such that the keys are of the form \"XF\" and \"XT\", such that nodes[\"XT\"] is the list of\n",
+    "        nodes representing the term (\"X\", True), and nodes[\"XT\"] is the list of nodes representing the term (\"X\", False) \n",
+    "        \"\"\"\n",
+    "        nodes = defaultdict(lambda : []) # set the default value of a key to an empty list\n",
+    "        \n",
```

Comment:
Let's take the opportunity to document a bit the rationale for this graph:
- if a node representing X is part of the MIS, then X __must__ be true;
- if a node representing NOT X is part of the MIS, then NOT X __must__ be true;
- otherwise, the term does not contribute to the solution (we don't care whether it's true or false).